clients/bigquery/storagewrite.go

```diff
@@ -0,0 +1,200 @@
+	case typing.Boolean.Kind:
+		fieldType = storagepb.TableFieldSchema_BOOL
+	case typing.Integer.Kind:
+		fieldType = storagepb.TableFieldSchema_INT64
+	case typing.Float.Kind:
+		fieldType = storagepb.TableFieldSchema_DOUBLE
+	case typing.EDecimal.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.String.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.ETime.Kind:
+		switch column.KindDetails.ExtendedTimeDetails.Type {
+		case ext.TimeKindType:
+			fieldType = storagepb.TableFieldSchema_TIME
+		case ext.DateKindType:
+			fieldType = storagepb.TableFieldSchema_DATE
+		case ext.DateTimeKindType:
+			fieldType = storagepb.TableFieldSchema_TIMESTAMP
+		default:
+			return nil, fmt.Errorf("unsupported extended time details type: %q", column.KindDetails.ExtendedTimeDetails.Type)
+		}
+	case typing.Struct.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+	case typing.Array.Kind:
+		fieldType = storagepb.TableFieldSchema_STRING
+		mode = storagepb.TableFieldSchema_REPEATED
+	default:
+		return nil, fmt.Errorf("unsupported column kind: %q", column.KindDetails.Kind)
+	}
+
+	return &storagepb.TableFieldSchema{
+		Name: column.Name(),
+		Type: fieldType,
+		Mode: mode,
+	}, nil
+}
+
+func columnsToMessageDescriptor(cols []columns.Column) (*protoreflect.MessageDescriptor, error) {
+	fields := make([]*storagepb.TableFieldSchema, len(cols))
+	for i, col := range cols {
+		field, err := columnToTableFieldSchema(col)
+		if err != nil {
+			return nil, err
+		}
+		fields[i] = field
+	}
+	tableSchema := storagepb.TableSchema{Fields: fields}
+
+	descriptor, err := adapt.StorageSchemaToProto2Descriptor(&tableSchema, "root")
+	if err != nil {
+		return nil, fmt.Errorf("failed to build proto descriptor: %w", err)
+	}
+	messageDescriptor, ok := descriptor.(protoreflect.MessageDescriptor)
+	if !ok {
+		return nil, fmt.Errorf("proto descriptor is not a message descriptor")
+	}
+	return &messageDescriptor, nil
+}
+
+// From https://cloud.google.com/java/docs/reference/google-cloud-bigquerystorage/latest/com.google.cloud.bigquery.storage.v1.CivilTimeEncoder
+// And https://cloud.google.com/pubsub/docs/bigquery#date_time_int
+func encodePacked64TimeMicros(value time.Time) int64 {
+	var result = int64(value.Nanosecond() / 1000)
+	result |= int64(value.Second()) << 20
+	result |= int64(value.Minute()) << 26
+	result |= int64(value.Hour()) << 32
+	return result
+}
+
+func rowToMessage(row map[string]any, columns []columns.Column, messageDescriptor protoreflect.MessageDescriptor, additionalDateFmts []string) (*dynamicpb.Message, error) {
+	message := dynamicpb.NewMessage(messageDescriptor)
+
+	for _, column := range columns {
+		field := message.Descriptor().Fields().ByTextName(column.Name())
+		if field == nil {
+			return nil, fmt.Errorf("failed to find a field named %q", column.Name())
+		}
+
+		value := row[column.Name()]
+
+		if value == nil {
+			continue
+		}
+
+		switch column.KindDetails.Kind {
```

Comment:
Can we pull this out into a separate function?

Reply:
And have it return the value so the caller can run `message.Set(field, value)`?

Reply:
I had that originally, and would have preferred to keep it that way, but the `typing.Array` type needs to be able to access the entire message: https://github.com/artie-labs/transfer/pull/715/files#diff-1340c750bdd06d343441e36dbe9c4877ca8fe7c9d09b2564227e40fe883f5710R190

Reply:
Got it